google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java

```diff
@@ -260,6 +262,84 @@ public void downloadTo(Path path) {
     downloadTo(path, new BlobSourceOption[0]);
   }
 
+  /**
+   * Uploads the given file path to this blob using specified blob write options.
+   *
+   * @param path file to be uploaded
+   * @param options blob write options
+   * @return updated blob
+   * @throws StorageException upon failure
+   */
+  public Blob uploadFrom(Path path, BlobWriteOption... options) {
+    if (!Files.exists(path)) {
+      throw new StorageException(0, path + ": No such file");
+    }
+    if (Files.isDirectory(path)) {
+      throw new StorageException(0, path + ": Is a directory");
+    }
+    try (InputStream input = Files.newInputStream(path)) {
+      return uploadFrom(input, options);
+    } catch (IOException e) {
+      throw new StorageException(e);
```

Comment:
IOException should be thrown directly here. It should not be converted to a runtime exception (and StorageException should probably not be a runtime exception)

Reply:
Thanks for looking at this PR.
I agree with all your comments, except ones related to IOException handling.
There are many places in the storage code where it is wraping to StorageException, that allows user to have a single handler for all kind of exceptions.  I'm not sure that IOException deserves a dedicated handler.

@frankyn  recommended extending Storage interface with upload functionality. I'm going to implement this in a separate PR.
Storage already has `create(BlobInfo, InputStream)` method which doesn't throw IOException, so adding `create(BlobInfo, Path)` which throws IOException in addition to StorageException will not look logically. I'm afraid people will write their own utility methods to wrap unnecessary exceptions.

Reply:
StorageException is a huge mistake. It should never have been a runtime exception. Consistency is not a sufficient reason to propagate this broken design to new code.

Reply:
Throwing IOException for a few methods will not fix the problem with existing design but make it only worse.
I agree,  BaseServiceException should be checked for better design. Does it mean that we cannot extend functionality unless https://github.com/googleapis/java-core/issues/57 is fixed?

Reply:
It won't fix the existing design, but it won't make the problem worse. Adding new methods that also make this mistake will make the problem worse.

Reply:
fixed